### PR TITLE
virtualization/amazon-init: enable option

### PR DESCRIPTION
**Pull request by urbas**
This is a non-functional, backwards-compatible change.

This change adds the `enable` option to the `amazon-init` module.

###### Motivation for this change

To make it possible to disable the `amazon-init` service at bootup.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [x] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by urbas** [comment]
I have used and verified this on a few EC2 instances.

@SuperSandro2000: May I ask for a suggestion on who to include as a reviewer for this module?

**Review by tomberek — APPROVED** [review]

File: nixos/modules/virtualisation/amazon-init.nix [hunk 1]
```diff
@@ -41,20 +45,45 @@ let
...
+      default = true;
+      type = types.bool;
+      description = ''
+        Enable or disable the amazon-init service.
+      '';
+    };
+
+    script = mkOption {
+      default = script;
+      type = types.str;
```
Comment (SuperSandro2000) [inline nixos/modules/virtualisation/amazon-init.nix:60]: I am really unsure about this type. Is it lines that get executed or a file? Maybe adding an example would be helpful, too.

Comment (tomberek) [reply nixos/modules/virtualisation/amazon-init.nix:60]: This is a minimal change compared to before. The text of the script is stored in the `script` variable. The default is current behavior.

Comment (urbas) [reply nixos/modules/virtualisation/amazon-init.nix]: @SuperSandro2000: You're right, the actual type is `lines`.

To dive deeper, this option is used in `systemd.services.amazon-init.script`, which is a string option defined in `nixos/modules/system/boot/systemd.nix` (and that is of type `lines`.

I changed this now and also added an example.

PTAL

**Review by SuperSandro2000 — COMMENTED** [review]
I think this is fine but I am not completely sure.

**Comment by mkaito** [comment]
Having an `enable` option is good.

As for the script override, I'd rather see us make the existing script better and more useful.

**Comment by urbas** [comment]
@mkaito: Very good point. Thinking more about it, the script override doesn't solve my problem, but a more powerful script would.

I have removed the script option from this PR and will propose an improvement to the script in another PR.

Please take another look.

**Review by mkaito — APPROVED** [review]
Adding the enable option is a good thing, as I said. Have a formal approval.

If the script isn't useful or good, then what's the point of having it in the first place. Here's a few thoughts I've had during our time relying on it at work.

I think the idea behind it is a good one (add some channels, deploy a nix expression), but the current implementation is somewhat lacking.

It might not be desirable to run `nixos-rebuild` every time the server boots, for various reasons.

Updating instance meta-data is cumbersome and can't be done while the instance is running, so updating the deployed nix expression is harder than it needs to be.

This unit should also arguably be an opt-in rather than a default when importing the `amazon-image` profile.

For your particular use case, you can just add an ad-hoc systemd unit that does whatever it is you need done. No need to mess with `amazon-init` other than disabling it.

And if you'll forgive a little shilling, if you happen to be looking for better ways to deploy server configurations, why not have a look at [deploy-rs](https://github.com/serokell/deploy-rs/)?

**Comment by urbas** [comment]
Thanks for the approve!

Regarding the "usefulness of the script": It's actually the only way to provision your instance at boot without building a custom AMI. I would very much like to avoid building my own AMI if possible. 

The script is essential for spot instances (which I'm using) and fleets where you have to rely on boot-time provisioning.

The way avoid running the script on every reboot is to disable the amazon-init service with the option I introduced in this PR (that's why this option is so useful).

I'll definitely  take a look at deploy-rs.

For the next step though I'm thinking of adding a minimal change to the existing script to support `nix-shell` scripts. I think this will be most useful, NixOS-idiomatic, and the simplest change to review.

**Comment by urbas** [comment]
@SuperSandro2000: I think this change is ready to go. Could you please take another look?

**Pull request merged**